src/Workspaces/Core/Portable/Utilities/AsyncLazy`1.cs

```diff
@@ -530,24 +544,36 @@ private void CompleteFromTaskSynchronouslyStub(object task)
 
             public void CompleteFromTaskSynchronously(Task<T> task)
             {
-                if (task.Status == TaskStatus.RanToCompletion)
+                // AsyncTaskMethodBuilder doesn't give us Try* methods, and the Set methods may throw if the task
+                // is already completed. The belief is that the race is somewhere between rare to impossible, and
+                // so we'll do a quick check to see if the task is already completed or otherwise just give it a shot
+                // and catch it if it fails
+                if (_taskBuilder.Task.IsCompleted)
                 {
-                    if (TrySetResult(task.Result))
-                    {
-                        _cancellationTokenRegistration.Dispose();
-                    }
+                    return;
                 }
-                else if (task.Status == TaskStatus.Faulted)
+
+                try
                 {
-                    if (TrySetException(task.Exception))
+                    if (task.Status == TaskStatus.RanToCompletion)
```

Comment:
Per my above misunderstanding of how this method is used, ignore this comment; I thought it was being called when the task had already completed.  But do consider caching Task.Status into a local rather than accessing it multiple times.